timemachine/cpp/src/kernels/chiral_utils.cuh

```diff
@@ -0,0 +1,189 @@
+
+template <typename RealType> __device__ struct Matrix { RealType x0, y0, z0, x1, y1, z1, x2, y2, z2; };
+
+template <typename RealType> __device__ struct Vector {
+
+    RealType x, y, z;
+
+    __device__ RealType dot(Vector<RealType> const &other) { return x * other.x + y * other.y + z * other.z; }
+
+    __device__ RealType norm() const { return std::sqrt(x * x + y * y + z * z); }
+
+    __device__ Vector<RealType> unit() const {
+        RealType n = norm();
+        return {x / n, y / n, z / n};
+    }
+
+    __device__ Matrix<RealType> unit_jac() const {
+        RealType n = this->norm();
+        Vector<RealType> v = this->unit();
+        return {
+            (1 - v.x * v.x) / n,
+            (-v.x * v.y) / n,
+            (-v.x * v.z) / n,
+            (-v.y * v.x) / n,
+            (1 - v.y * v.y) / n,
+            (-v.y * v.z) / n,
+            (-v.z * v.x) / n,
+            (-v.z * v.y) / n,
+            (1 - v.z * v.z) / n};
+    }
+
+    __device__ Vector<RealType> operator+(Vector<RealType> const &obj) const {
+        return {x + obj.x, y + obj.y, z + obj.z};
+    }
+
+    __device__ Vector<RealType> operator-(Vector<RealType> const &obj) const {
+        return {x - obj.x, y - obj.y, z - obj.z};
+    }
+
+    // unary minus
+    __device__ Vector<RealType> operator-() const { return {-x, -y, -z}; }
+};
+
+template <typename RealType> __device__ Vector<RealType> mul_vm(Vector<RealType> v, Matrix<RealType> m) {
```

Comment:
nit: would it be more conventional to have the matrix as the left argument? (another idea: could be a method of `Matrix`, so you could do `mat.dot(vec)`)

Reply:
this is a left multiply of a vector * matrix - (note the signature of `_vm`) so probably best to be consistent in its args?

Reply:
Got it, makes sense. I misread the implementation as multiplication of a matrix by a column vector on the right.